File: src/main/java/com/hamosad1657/lib/sensors/HaColorSensor.java

package com.hamosad1657.lib.sensors;

import edu.wpi.first.util.sendable.Sendable;
import edu.wpi.first.util.sendable.SendableBuilder;
import edu.wpi.first.wpilibj.I2C;
import edu.wpi.first.wpilibj.util.Color;
import com.revrobotics.ColorSensorV3;
import com.revrobotics.ColorSensorV3.RawColor;

public class HaColorSensor implements Sendable {
	private final ColorSensorV3 colorSensor;

	private final int minCm = 1;
	private final int maxCm = 10;
	private final int minProx = 0;
	private final int maxProx = 2047;

	/**
	 * @param port - The I2C port that the sensor is connected to.
	 */
	public HaColorSensor(I2C.Port port) {
		this.colorSensor = new ColorSensorV3(port);
	}

	/**
	 * @param minColor - The minimum acceptable color values (inclusive).
	 * @param maxColor - The maximum acceptable color values (inclusive).
	 * 
	 * @return Whether the detected color is in the specified range (inclusive).
	 */
	public boolean isColorInRange(RawColor minColor, RawColor maxColor) {
		RawColor color = colorSensor.getRawColor();
		return (color.red >= minColor.red && color.red <= maxColor.red && color.blue >= minColor.blue
				&& color.blue <= maxColor.blue && color.green >= minColor.green && color.green <= maxColor.green);
	}

	/**
	 * @param minProximity - The minimum acceptable proximity (inclusive).
	 * @param maxProximity - The maximum acceptable proximity (inclusive).
	 * 
	 * @return Whether the proximity is in the specified range (inclusive).
	 */
	public boolean isObjectInProximityRange(double minProximity, double maxProximity) {
		int proximity = this.getProximity();
		return proximity >= minProximity && proximity <= maxProximity;
	}

	/**
	 * @return The raw color detected by the sensor, as a RawColor object.
	 */
	public RawColor getColor() {
		return this.colorSensor.getRawColor();
	}

	/**
	 * 
	 * @return The red value of the detected color.
	 */
	public double getRed() {
		return this.colorSensor.getRed();
	}

	/**
	 * 
	 * @return The green value of the detected color.
	 */
	public double getGreen() {
		return this.colorSensor.getGreen();
	}

	/**
	 * 
	 * @return The blue value of the detected color.
	 */
	public double getBlue() {
		return this.colorSensor.getBlue();
	}

	/**
	 * 
	 * @return The IR color value detected by the sensor (in CIE 1931 XYZ colorspace).
	 */
	public int getIR() {
		return this.colorSensor.getIR();
	}

	/**
	 * 
	 * @return The proximity value of the sensor ranging from 0 (object is close) to 2047 (object is far away).
	 */
	public int getProximity() {
		return this.maxProx - this.colorSensor.getProximity(); // Flip the range from [2047, 0] to [0, 2047]
	}

	/**
	 * 
	 * @return The proximity value of the sensor in cm
	 */
	public int getCmProximity() {
		return this.minCm
				+ ((this.getProximity() - this.minProx) * (this.maxCm - this.minCm)) / (this.maxProx - this.minProx);
	}

	/**
	 * @param minColor - The minimum acceptable color values (inclusive).
	 * @param maxColor - The maximum acceptable color values (inclusive).
	 * 
	 * @return Whether the detected color is in the specified range (inclusive).
	 */
	public boolean isColorInRangePercent(Color minColor, Color maxColor) {
		Color color = this.colorSensor.getColor();
		return (color.red >= minColor.red && color.red <= maxColor.red && color.blue >= minColor.blue
				&& color.blue <= maxColor.blue && color.green >= minColor.green && color.green <= maxColor.green);
	}

	/**
	 * @return The raw color detected by the sensor, as a RawColor object.
	 */
	public Color getColorPercent() {
		return this.colorSensor.getColor();
	}

	/**
	 * 
	 * @return The red value of the detected color.
	 */
	public double getRedPercent() {
		return this.colorSensor.getColor().red;
	}

	/**
	 * 
	 * @return The green value of the detected color.
	 */
	public double getGreenPercent() {
		return this.colorSensor.getColor().green;
	}

	/**
	 * 
	 * @return The blue value of the detected color.
	 */
	public double getBluePercent() {
		return this.colorSensor.getColor().blue;
	}

	@Override
	public void initSendable(SendableBuilder builder) {
		builder.setSmartDashboardType("HaColorSensor");

		builder.addDoubleProperty("Red", this::getRed, null);
		builder.addDoubleProperty("Green", this::getGreen, null);
		builder.addDoubleProperty("Blue", this::getBlue, null);

		builder.addDoubleProperty("Red %", this::getRedPercent, null);
		builder.addDoubleProperty("Green %", this::getGreenPercent, null);
		builder.addDoubleProperty("Blue %", this::getBluePercent, null);

		builder.addDoubleProperty("Proximity (CM)", this::getCmProximity, null);

		builder.addDoubleProperty("Proximity (0-2047)", this::getProximity, null);
		builder.addDoubleProperty("IR", this::getIR, null);
	}
}

File: src/main/kotlin/com/hamosad1657/lib/sensors/HaColorSensor.kt
package com.hamosad1657.lib.sensors

import com.revrobotics.ColorSensorV3
import com.revrobotics.ColorSensorV3.RawColor
import edu.wpi.first.util.sendable.Sendable
import edu.wpi.first.util.sendable.SendableBuilder
import edu.wpi.first.wpilibj.I2C
import edu.wpi.first.wpilibj.util.Color

class HaColorSensor(port: I2C.Port?) : Sendable {
    private val colorSensor: ColorSensorV3
    private val maxProx = 2047

    init {
        colorSensor = ColorSensorV3(port)
    }

    /**
     * @param minColor - The minimum acceptable color values (inclusive).
     * @param maxColor - The maximum acceptable color values (inclusive).
     *
     * @return Whether the detected color is in the specified range (inclusive).
     */
    fun isColorInRange(minColor: RawColor, maxColor: RawColor): Boolean {
        val color = colorSensor.rawColor
        return color.red >= minColor.red && color.red <= maxColor.red && color.blue >= minColor.blue && color.blue <= maxColor.blue && color.green >= minColor.green && color.green <= maxColor.green
    }

    /**
     * @param minProximity - The minimum acceptable proximity (inclusive).
     * @param maxProximity - The maximum acceptable proximity (inclusive).
     *
     * @return Whether the proximity is in the specified range (inclusive).
     */
    fun isObjectInProximityRange(minProximity: Double, maxProximity: Double): Boolean {
        val proximity = proximity
        return proximity >= minProximity && proximity <= maxProximity
    }

    val color: RawColor
        /**
         * @return The raw color detected by the sensor, as a RawColor object.
         */
        get() = colorSensor.rawColor
    private val red: Double
        /**
         *
         * @return The red value of the detected color.
         */
        get() = colorSensor.red.toDouble()
    private val green: Double
        /**
         *
         * @return The green value of the detected color.
         */
        get() = colorSensor.green.toDouble()
    private val blue: Double
        /**
         *
         * @return The blue value of the detected color.
         */
        get() = colorSensor.blue.toDouble()
    private val IR: Int
        /**
         *
         * @return The IR color value detected by the sensor (in CIE 1931 XYZ colorspace).
         */
        get() = colorSensor.ir
    private val proximity: Int
        /**
         *
         * @return The proximity value of the sensor ranging from 0 (object is close) to 2047 (object is far away).
         */
        get() = maxProx - colorSensor.proximity // Flip the range from [2047, 0] to [0, 2047]
    private val cmProximity: Int
        /**
         *
         * @return The proximity value of the sensor in cm
         */
        get() {
            val minCm = 1
            val maxCm = 10
            val minProx = 0
            return (minCm
                    + (proximity - minProx) * (maxCm - minCm) / (maxProx - minProx))
        }

    /**
     * @param minColor - The minimum acceptable color values (inclusive).
     * @param maxColor - The maximum acceptable color values (inclusive).
     *
     * @return Whether the detected color is in the specified range (inclusive).
     */
    fun isColorInRangePercent(minColor: Color, maxColor: Color): Boolean {
        val color = colorSensor.color
        return color.red >= minColor.red && color.red <= maxColor.red && color.blue >= minColor.blue && color.blue <= maxColor.blue && color.green >= minColor.green && color.green <= maxColor.green
    }

    val colorPercent: Color
        /**
         * @return The raw color detected by the sensor, as a RawColor object.
         */
        get() = colorSensor.color
    private val redPercent: Double
        /**
         *
         * @return The red value of the detected color.
         */
        get() = colorSensor.color.red
    private val greenPercent: Double
        /**
         *
         * @return The green value of the detected color.
         */
        get() = colorSensor.color.green
    private val bluePercent: Double
        /**
         *
         * @return The blue value of the detected color.
         */
        get() = colorSensor.color.blue

    override fun initSendable(builder: SendableBuilder) {
        builder.setSmartDashboardType("HaColorSensor")
        builder.addDoubleProperty("Red", { red }, null)
        builder.addDoubleProperty("Green", { green }, null)
        builder.addDoubleProperty("Blue", { blue }, null)
        builder.addDoubleProperty("Red %", { redPercent }, null)
        builder.addDoubleProperty("Green %", { greenPercent }, null)
        builder.addDoubleProperty("Blue %", { bluePercent }, null)
        builder.addDoubleProperty("Proximity (CM)", { cmProximity.toDouble() }, null)
        builder.addDoubleProperty("Proximity (0-2047)", { proximity.toDouble() }, null)
        builder.addDoubleProperty("IR", { IR.toDouble() }, null)
    }
}
